packages/beacon-node/src/chain/chain.ts

```diff
@@ -991,4 +992,11 @@ export class BeaconChain implements IBeaconChain {
       }
     }
   }
+
+  async getBlockRewards(block: allForks.FullOrBlindedBeaconBlock): Promise<BlockRewards> {
+    const preState = (await this.regen.getPreState(block, {dontTransferCache: true}, RegenCaller.restApi)).clone();
```

Comment:
> A little sketchy that this can trigger a regen.
What's your concern in particular? I don't think `getPreState` modifies anything under the hood. To get the preState of a block I don't see a way without calling regen. At some point `regen.getState()` must be called.

> Is this the only non-debug endpoint that can do that?
This is the only way that I know of cc. @tuyennhv

Reply:
> Is this the only non-debug endpoint that can do that?

It looks like this is the only one, can be confirmed by searching for `allowRegen` which is only set to true for debug state apis

> What's your concern in particular?

The main concern is probably that you can use this API to easily DoS our public nodes as regen is quite expensive.

What about adding rewards APIs as their own namespace and disabling them by default?

e.g. we also have light client APIs on their own namespace even though those are part of /beacon as per spec
https://github.com/ChainSafe/lodestar/blob/347c95fb0da219b5a4d551b7ce014f2ceda7ae48/packages/api/src/beacon/server/index.ts#L41

we could then simply not enable rewards APIs by default here
https://github.com/ChainSafe/lodestar/blob/347c95fb0da219b5a4d551b7ce014f2ceda7ae48/packages/beacon-node/src/api/rest/index.ts#L19

Reply:
> What about adding rewards APIs as their own namespace and disabling them by default?

I read the documentation other CL clients and looks like everyone has rewards endpoint enabled by default. Don’t know if the users expect this endpoint would be enabled in the public nodes. 

Alternatively, we can limit the queried block from last finalized checkpoint to head similar to what Teku behaves under `prune` mode: https://docs.teku.consensys.io/how-to/use-rewards-api#limitations . This way we only need to check `stateCache` and `checkpointStateCache` without triggering a regen request

Reply:
As an aside, It may be worth thinking deeper about our strategy around serving more expensive queries.

Should we support expensive queries? Behind a feature flag or flags? If so, what does that look like?
Behind extra namespaces for each additional feature-set in the APIs?
What does the architecture of generating expensive data look like? Queue that deprioritizes non-urgent work? Separate worker that handles expensive queries? Other?

Reply:
yes this is really a concern because with `PersistentCheckpointStateCache`, using `getPreState` means it will reload checkpoint state if needed

> We could add and use a regen.getPreStateSync that acts like regen.getStateSync in that it only checks cached state values and returns undefined if the state isn't cached. IMO this is a good compromise, since it allows us to serve the data in most cases, but doesn't open us up to any DoS issues.

in this specific scenario, if we can get a cached pre state, we can also get post state and get the cached reward from post state, so it's no use to have `getPreStateSync()`

I think we only want to support getting cached reward in post state for lodestar (it means it'll work for 64-96 blocks for now). If demand arises from any specific use case, we can enhance later

Reply:
> in this specific scenario, if we can get a cached pre state, we can also get post state and get the cached reward from post state, so it's no use to have getPreStateSync()

For the purpose of this endpoint, I echo with @wemeetagain 's idea to only serve blocks which its corresponding postState is cached. 
So we should just drop the preState parameter in `computeBlockRewards(block, preState, postState)`. 

> so it's no use to have `getPreStateSync()`

The other two rewards endpoint (attestation and sync committee) both require `preState` so we will need to have `getPreStateSync()` implemented at some point. We can avoid `getPreStateSync()` for this endpoint because we have block rewards cached in `BeaconStateCache.proposerRewards: RewardCache` but that's not the case for the other endpoints.